app/controllers/gazetteers_controller.rb

```diff
@@ -0,0 +1,224 @@
+        @recent_objects = Gazetteer
+          .recent_from_project_id(sessions_current_project_id)
+          .order(updated_at: :desc)
+          .limit(10)
+        render '/shared/data/all/index'
+      end
+      format.json do
+        return
+        # no filter on GZs yet
+      end
+    end
+  end
+
+  # GET /gazetteers/1 or /gazetteers/1.json
+  def show
+  end
+
+  # GET /gazetteers/new
+  def new
+    respond_to do |format|
+      format.html { redirect_to new_gazetteer_task_path }
+    end
+  end
+
+  # GET /gazetteers/1/edit
+  def edit
+    respond_to do |format|
+      format.html {
+        redirect_to new_gazetteer_task_path gazetteer_id: @gazetteer.id
+      }
+    end
+  end
+
+  # GET /gazetteers/list
+  def list
+    @gazetteers = Gazetteer
+      .with_project_id(sessions_current_project_id)
+      .page(params[:page]).per(params[:per])
+  end
+
+  # POST /gazetteers.json
+  def create
+    @gazetteer = Gazetteer.new(gazetteer_params)
+
+    @gazetteer.build_gi_from_shapes(shape_params['shapes'])
+    if @gazetteer.errors.include?(:base)
+      render json: @gazetteer.errors, status: :unprocessable_entity
+      return
+    end
+
+    begin
+      Gazetteer
+        .save_and_clone_to_projects(@gazetteer, projects_param['projects'])
+
+      render :show, status: :created, location: @gazetteer
+    rescue ActiveRecord::RecordInvalid => e
+      render json: { errors: e.message }, status: :unprocessable_entity
+    end
+  end
+
+  # PATCH/PUT /gazetteers/1
+  # PATCH/PUT /gazetteers/1.json
+  def update
+    respond_to do |format|
+      if @gazetteer.update(gazetteer_params)
+        format.html { redirect_to gazetteer_url(@gazetteer) }
+        format.json { render :show, status: :ok }
+      else
+        format.html { render :edit, status: :unprocessable_entity }
+        format.json { render json: @gazetteer.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  # DELETE /gazetteers/1
+  # DELETE /gazetteers/1.json
+  def destroy
+    @gazetteer.destroy!
+
+    respond_to do |format|
+      format.html {
+        redirect_to gazetteers_url,
+          notice: 'Gazetteer was successfully destroyed.'
+      }
+      format.json { head :no_content }
+    end
+  end
+
+  def autocomplete
+    @gazetteers = ::Queries::Gazetteer::Autocomplete.new(
+      params.require(:term),
+      project_id: sessions_current_project_id,
+    ).autocomplete
+  end
+
+  def search
+    if params[:id].blank?
+      redirect_to(gazetteer_path,
+                  alert: 'You must select an item from the list with a click or tab press before clicking show.')
+    else
+      redirect_to gazetteer_path(params[:id])
+    end
+  end
+
+  # GET /gazetteers/download
+  def download
+    send_data Export::CSV.generate_csv(
+        Gazetteer.where(project_id: sessions_current_project_id)
+      ),
+      type: 'text',
+      filename: "gazetteers_#{DateTime.now}.tsv"
+  end
+
+  # POST /gazetteers/import.json
+  def import
```

Comment:
@kleintom A little too much logic in here I think,  Can we move almost everything to a model or lib?